sys/vm_map.c

```diff
@@ -91,22 +94,31 @@ static bool vm_map_insert_entry(vm_map_t *vm_map, vm_map_entry_t *entry) {
 
 vm_map_entry_t *vm_map_find_entry(vm_map_t *vm_map, vm_addr_t vaddr) {
   vm_map_entry_t *etr_it;
-  TAILQ_FOREACH (etr_it, &vm_map->list, map_list)
-    if (etr_it->start <= vaddr && vaddr < etr_it->end)
+  rw_enter(&vm_map->rwlock, RW_READER);
+  TAILQ_FOREACH (etr_it, &vm_map->list, map_list) {
+    if (etr_it->start <= vaddr && vaddr < etr_it->end) {
+      rw_leave(&vm_map->rwlock);
       return etr_it;
+    }
+  }
+  rw_leave(&vm_map->rwlock);
   return NULL;
 }
 
 static void vm_map_remove_entry(vm_map_t *vm_map, vm_map_entry_t *entry) {
+  rw_assert(&vm_map->rwlock, RW_WLOCKED);
   vm_map->nentries--;
   vm_object_free(entry->object);
   TAILQ_REMOVE(&vm_map->list, entry, map_list);
   kfree(mpool, entry);
 }
 
 void vm_map_delete(vm_map_t *map) {
+  rw_enter(&map->rwlock, RW_WRITER);
```

Comment:
The first variant would look cool, but I don't think there is any way to implement this, as the macro would not control neither the beginning nor the end of the scope.

The second style could be improved even a bit more:

```c
void rw_guard_leave(rwlock_t **rwlock) {
  /* Dereference required... */
  rw_leave(*rwlock);
}

#define rw_lock_guard(prwlock, mode)                                           \
  rwlock_t *rw_lock_guard_##__LINE__                                           \
    __attribute__((cleanup(rw_guard_leave))) = prwlock;                        \
  rw_enter(prwlock, mode);
```

and would be used with bare:
```c
rw_lock_guard(&map->rwlock, RW_WRITER);
```

We can also have a similar macro for `mtx_t`.